notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,526 @@
+
+This notebook builds on the [denoising autoencoder example](mnist_autoencoder_using_kino.livemd) and turns the simple autoencoder into a variational one for the same dataset.
+
+## Training a simple autoencoder
+
+This section will proceed without much explanation as most of it is extracted from [denoising autoencoder example](mnist_autoencoder_using_kino.livemd). If anything here doesn't make sense, take a look at that notebook for an explanation.
+
+```elixir
+defmodule Data do
+  @moduledoc """
+  A module to hold useful data processing utilities, 
+  mostly extracted from the previous notebook
+  """
+
+  @doc """
+  Converts the given image into a `Kino.Image`.
+
+  `image` must be a single channel `Nx` tensor with pixel values between 0 and 1.
+  `height` and `width` are the output size in pixels
+  """
+  def image_to_kino(image, height \\ 200, width \\ 200) do
+    image
+    |> Nx.multiply(255)
+    |> Nx.as_type(:u8)
+    |> Nx.transpose(axes: [:height, :width, :channels])
+    |> StbImage.from_nx()
+    |> StbImage.resize(height, width)
+    |> StbImage.to_binary(:png)
+    |> Kino.Image.new(:png)
+  end
+
+  @doc """
+  Converts image data from `Scidata.MNIST` into an `Nx` tensor and normalizes it.
+  """
+  def preprocess_data(data) do
+    {image_data, _labels} = data
+    {images_binary, type, shape} = image_data
+
+    images_binary
+    |> Nx.from_binary(type)
+    # Since pixels are organized row-wise, reshape into rows x columns
+    |> Nx.reshape(shape, names: [:images, :channels, :height, :width])
+    # Normalize the pixel values to be between 0 and 1
+    |> Nx.divide(255)
+  end
+
+  @doc """
+  Converts a tensor of images into paired, batched images for model training
+  """
+  def prepare_training_data(images, batch_size) do
+    batched_images = images |> Nx.to_batched(batch_size)
+    Stream.zip(batched_images, batched_images)
+  end
+end
+```
+
+```elixir
+train_images = Data.preprocess_data(Scidata.FashionMNIST.download())
+test_images = Data.preprocess_data(Scidata.FashionMNIST.download_test())
+
+Kino.render(train_images[[images: 0]] |> Data.image_to_kino())
+Kino.render(test_images[[images: 0]] |> Data.image_to_kino())
+
+:ok
+```
+
+Now for our simple autoencoder model. We won't be using a denoising autoencoder here.
+
+Note that we're giving each of the layers a name - the reason for this will be apparent later.
+
+I'm also using a small custom layer to shift and scale the output of the sigmoid layer slightly so it can hit the 0 and 1 targets. I noticed the gradients tend to explode without this.
+
+```elixir
+defmodule CustomLayer do
+  import Nx.Defn
+
+  def scaling_layer(%Axon{} = input, _opts \\ []) do
+    Axon.layer(&scaling_layer_impl/2, [input])
+  end
+
+  defnp scaling_layer_impl(x, _opts \\ []) do
+    x
+    |> Nx.subtract(0.05)
+    |> Nx.multiply(1.2)
```

Comment:
This is interesting, so this got rid of NaNs...where did these values come from?

Reply:
These values were just made up to get the sigmoid to cover 0 and 1 (I think this is a similar trick to what fast.ai uses). Based on my understanding of `layer_norm` I don't think it'll enable the sigmoid to reach 1.0 as an output, is that right?

The issue seemed to be that since the mnist values are often actually 0 and 1 and sigmoid couldn't get there, as the network converged it droves certain params to more extremes to try to reach those values.